nengo_spa/compiler/ast.py

```diff
@@ -0,0 +1,288 @@
+        return self.sink.obj, SpaNetwork.get_input_vocab(self.sink.obj)
+
+
+class ActionsScope(object):
+    """Captures the scope that names in action rules refer to."""
+
+    builtins = {
+        'dot': DotProduct,
+        'reinterpret': Reinterpret,
+        'translate': Translate}
+
+    def __init__(self, stacklevel=1):
+        frame = inspect.currentframe()
+        for _ in range(stacklevel):
+            frame = frame.f_back
+        self.locals = frame.f_locals
+        self.globals = frame.f_globals
+        self.py_builtins = frame.f_builtins
+
+    def __getitem__(self, key):
+        if key == '__tracebackhide__':  # gives better tracebacks in py.test
+            item = False
+        elif key in self.builtins:
+            item = self.builtins[key]
+        elif key in self.locals:
+            item = self.locals[key]
+        elif key in self.globals:
+            item = self.globals[key]
+        elif key in self.py_builtins:
+            item = self.py_builtins[key]
+        elif key[0].isupper():
+            item = Symbol(key)
+        else:
+            raise SpaNameError(key, 'name')
+
+        if isinstance(item, SpaNetwork):
+            item = Module(key, item)
+        return item
+
+
+class AstBuilder(object):
+    """Builds the AST from a parse tree.
+
+    Parameters
+    ----------
+    scope : ActionsScope, optional
+        Used to resolve names in the action rules. If not given, it defaults to
+        the scope of the calling function.
+    """
+
+    def __init__(self, scope=None):
+        if scope is None:
+            scope = ActionsScope(stacklevel=2)
+        self.scope = scope
+        self._channeled = False
+        self._encoding = None
+        self._name = None
+        self._index = 0
+
+    def build(self, parse_tree):
+        """Builds a parse tree (or subtree) into an AST.
+
+        Returns
+        -------
+        list
+            List of built top-level AST nodes.
+        """
+        ast_nodes = []
+        for child in parse_tree:
+            if tokens.is_token_info(child):
+                ast_nodes.extend(self.build_token(child))
+            else:
+                build_fn_name = 'build_' + child[0]
+                if hasattr(self, build_fn_name):
+                    ast_nodes.extend(getattr(self, build_fn_name)(child[1]))
+                else:
+                    ast_nodes.extend(self.build(child[1]))
+        return ast_nodes
+
+    def build_token(self, token):
+        if token[0] == tokens.ENCODING:
+            self._encoding = token.string
+        return []
+
+    def build_expr(self, parse_tree):
+        expr = self._untokenize_expr(parse_tree)
+        return [eval(expr, {}, self.scope)]
+
+    def build_sink(self, parse_tree):
+        expr = self._untokenize_expr(parse_tree)
+        obj = eval(expr, dict(self.scope.locals), dict(self.scope.globals))
+        return [Sink(expr.strip(), obj)]
+
+    def _untokenize_expr(self, parse_tree):
```

Comment:
This method expects the parse tree of the `expr` rule in the grammar (which actually won't be a tree, but just a flat list of tokens). So `parse_tree` represents a single expression (or what the SPA grammar defines as such, it might differ from the exact definiton of a Python expression). The parse tree (i.e. the list of tokens) is untokenized, that is, it is converted back into a string which is then returned.